fix: fix field alignment to avoid panics on 32bit OSes

hairyhenderson opened the pull request:
Fixes #43

To solve this, I:
- enabled the `fieldalignment` linter (which is a sub-linter of `govet`)
- ran the `fieldalignment -fix` command to fix misaligned structs that were non-obvious
- switch from using e.g. `int64` + `atomic.StoreInt64` to using `atomic.Int64` (etc) and because of this...
- drop Go 1.18 support, and require Go 1.19+ (where the new atomic types were introduced)
- added a new CI job to test in a 32-bit container (so far just Intel 386, which should catch the same bugs as arm32/etc)
- added some new tests to cover `keyvalue/blob.Bytes`, which was what was giving me trouble initially

hairyhenderson commented on keyvalue/blob/bytes.go:
```diff
@@ -16,14 +16,14 @@ var (
 		GrowBlob
 		TruncateBlob
-	} = &Bytes{}
+	} = (*Bytes)(nil)
```
This change isn't related, but is a (very mild) refactoring that lets the compiler remove this (I think we save 8 bytes of memory) - it's a little more idiomatic to do it this way.

JohnStarich commented on keyvalue/file.go:
```diff
@@ -33,12 +33,12 @@ type file struct {
+	fs   *FS
+	path string // path is stored as the "key", keeping it here is for generating hackpadfs.FileInfo's
+
 	modeOverride    *hackpadfs.FileMode
```
A tiny nit: Since the ordering isn't as flexible anymore and gofmt likes to align the names/types, let's delete the blank lines between fields for all of the realigned structs.

The runOnceFileRecord one in particular is a bit weird. In a future PR, I imagine it'd be worth a small refactor to group the sync.Once and value in a common type. Go 1.21's `sync.OnceValue` looks promising.

hairyhenderson replied:
> Since the ordering isn't as flexible anymore [...]

To be clear, the ordering was never flexible - alignment has always mattered...

I had the same thought about `sync.OnceValue`, but I didn't want to update Go too far, partly because the linter catches more things and I've spent way too much time on this already... A future PR sounds good.

I'll delete blank lines.

JohnStarich replied:
I hear ya, I was referring to the linter gating things now. Less flexible in the eyes of CI, if you will.

JohnStarich commented on keyvalue/record.go:
```diff
@@ -112,34 +112,34 @@ func (b *BaseFileRecord) Sys() interface{} {
-
-	mode     hackpadfs.FileMode
-	modeOnce sync.Once
+	dataDone atomic.Int64
```
If I remember right, dataDone behaves more like a `bool` but the atomic package didn't have those operations available at the time. Let's try to use `atomic.Bool` here.

hairyhenderson replied:
worth it, sure.

JohnStarich commented on mount_test.go:
```diff
@@ -13,9 +13,9 @@ func TestStripErrPathPrefix(t *testing.T) {
+		expectErr    error
 		name         string
 		mountSubPath string
-		expectErr    error
```
Same as fstest, if feasible let's avoid reordering the table driven test case types for better readability.

JohnStarich commented on tar/bufferpool.go:
```diff
@@ -7,40 +7,41 @@ import (
 func (p *bufferPool) addBuffer() {
 	for {
-		count := atomic.LoadInt64(&p.count)
+		// count := atomic.LoadInt64(&p.count)
```
Oops, stray comment?
```suggestion
```

JohnStarich reviewed (changes requested):
I'm not very experienced with these kinds of alignment issues, so I'm definitely open to a bit of discussion on how to handle these.

If at all possible, my strong preference is to maintain field ordering in the table driven test case types to keep it easier to quickly read and understand the tests. Of course if the tests fail, we'll need to do something else :thinking:

hairyhenderson commented:
> If at all possible, my strong preference is to maintain field ordering in the table driven test case types

I can revert the test files, no big deal.

hairyhenderson commented on .github/workflows/ci.yml:
```diff
@@ -40,3 +39,13 @@ jobs:
         COVERALLS_TOKEN: ${{ secrets.GITHUB_TOKEN }}
         COVERAGE_VERSION: 1.19
+  test32bit:
+    runs-on: ubuntu-latest
```
GitHub Actions doesn't provide 32-bit runners so the most reliable way to run as 32-bit is to use QEMU and Docker.

But, if this can be reproduced on a 64-bit runner simply running with GOARCH=386 that'd be fine. I'll mess with that and see if it works.

hairyhenderson replied:
ok, it was a bit more than setting `GOARCH` - I needed to tell `setup-go` to install the 386 version, but it works:

```
--- FAIL: TestChmod (0.00s)
panic: unaligned 64-bit atomic operation [recovered]
	panic: unaligned 64-bit atomic operation
```

JohnStarich replied:
Very interesting, I wonder why the env var was insufficient :thinking: On my Linux machine it panicked as expected.

hairyhenderson commented on fstest/file.go:
```diff
@@ -94,12 +94,12 @@ func TestFileReadAt(tb testing.TB, o FSOptions) {
+		bufSize     int
 		expectN     int
-		expectBuf   string
-		expectErr   error
```
If changing field ordering causes a struct to be unreadable, then I think we need some comments instead 😉

But, I'll revert and ignore these changes.

JohnStarich replied:
:wink: 

It's not that it became unreadable per se, but a tad less. Thank you.

hairyhenderson commented:
@JohnStarich thanks for the feedback, I think I've addressed it all - please have another look!

(I'm not sure why coveralls thinks the coverage decreased - I added new tests... ¯\\\_(ツ)_/¯)

JohnStarich commented on .github/workflows/ci.yml:
```diff
@@ -19,24 +19,24 @@ jobs:
+          - platform: macos-latest
+            goarch: '386'
+          - platform: windows-latest
+            goarch: '386'
```
This is great, thanks!

We may want to reduce the matrix a bit to a single 32 bit runner in the future, assuming we can prove the effective coverage is the same. As someone who's worked with 32 bit arch's more than me, do you foresee any problems with only validating those arch's on a single platform and Go version?

hairyhenderson replied:
The only reason I didn't also add 32-bit ARM as well is that it's a huge pain to do (without Docker) in GitHub Actions, because the GitHub-hosted runners are all AMD64 currently (ARM64 is coming this fall maybe?).

Another option is to use a self-hosted runner, and I happen to be deep in that rabbit-hole right now (https://github.com/hairyhenderson/gomplate/pull/2126) and it's going about as well as you'd expect (i.e. not at all well).

For this particular class of bug any 32-bit platform will generally do, but there are other classes of bugs that can crop up only on 32-bit ARM CPUs. Somewhat ironically it's extremely rare to run into 32-bit Intel systems except for niche embedded use-cases, but it's still quite common to run into 32-bit ARM given the massive proliferation of older Raspberry Pi-type systems and routers strewn across the IoT landscape.

JohnStarich reviewed (approved):
Thanks again! Super appreciate the thought and time you put into this.

I did have one question about the future of maintaining 32 bit support here, since I'm not as well versed in field alignment issues like this. https://github.com/hack-pad/hackpadfs/pull/44#discussion_r1650145979

hairyhenderson commented on fstest/file.go:
```diff
@@ -653,6 +654,7 @@ func TestFileSync(tb testing.TB, o FSOptions) {
 	assert.NoError(tb, file.Close())
 }
 
+//nolint:govet
```
It's at least awkward in the golangci config, if not impossible - it lets you disable whole linters (i.e. `govet` in this case), but not sub-linters like `fieldalignment`. 

If you want to avoid linting all tests with `govet`, then that's an option, but...

hairyhenderson commented on Makefile:
```diff
@@ -2,6 +2,7 @@ BROWSERTEST_VERSION = v0.7
 LINT_VERSION = 1.50.1
 GO_BIN = $(shell printf '%s/bin' "$$(go env GOPATH)")
 SHELL = bash
+ENABLE_RACE = true
```
🤦‍♂️ good catch - that can be deleted

hairyhenderson commented on Makefile:
```diff
@@ -31,14 +32,25 @@ test-deps:
+TEST_ARGS=-race
+# the race detector needs cgo (at least on Linux and Windows, and macOS until Go 1.20)
+export CGO_ENABLED=1
+endif
```
I was already in the midst doing a similar thing for gomplate so it at least wasn't a _new_ kind of pain 😅

It probably isn't always true, but races tend to happen in the same way on all platforms, so I wouldn't be too worried here.
